### ISSUE TITLE
Access container from migration

### ISSUE THREAD

**Issue by MekDrop** [lulco/phoenix#206]
I need to get table prefix for name and to do that I need somehow access services container. How to do that? With Doctrine migration that would look like this -> https://gist.github.com/chalasr/dc5cd628c9d16e255d4e2ea7b2f36b72

**Comment by lulco** [comment]
Hi, this is not supported natively. But you probably can require file which builds the container, or maybe create some static method in application / framework to get it to migration.

For example in Nette framework I can do this:

file: app/bootstrap.php
```php
$configurator = new Configurator();
$container = $configurator->createContainer();
return $container;
```

and in migration file: app/migration/FirstMigration.php
```php
class FirstMigration extends AbstractMigration
{
    protected function up(): void
    {
        $container = require __DIR__ . '/../boostrap.php';
    }
}
```

**Comment by MekDrop** [comment]
I'm using right now as global var but I'm thinking if there is a any better way.

Probably yet another working solution for me would be if there is a possibility to reuse existing other PDO connection because it's contains already existing method for prefixing table. 

Probably that in config file could look like:
```php
<?php

return [
    'migration_dirs' => [
        'phoenix' => __DIR__ . '/../testing_migrations/phoenix',
    ],
    'environments' => [
        'mysql' => new DatabaseConnection();
    ],
];
```

From the phoenix code I see that would be possible in `PdoAdapter` but it isn't possible without modifying `EnvironmentConfig` and `AdapterFactory` code and add some method in `PdoAdapter` like `getOriginalAdapter` for accessing original PDO adapter. I think this could be enough for solving this issue.

What do you think about such approach? If it good enough, I could try to make a new pull request for that.

**Comment by lulco** [comment]
Is your DatabaseConnection instance of PDO? I can imagine that EnvironmentConfig would return prepared PDO to AdapterFactory instead of this row:
```
$pdo = new PDO($config->getDsn(), $config->getUsername(), $config->getPassword());
```
would be:
```
$pdo = $config->getPdo() ?: new PDO($config->getDsn(), $config->getUsername(), $config->getPassword());
```

Config would look like:

```
<?php

return [
    'migration_dirs' => [
        'phoenix' => __DIR__ . '/../testing_migrations/phoenix',
    ],
    'environments' => [
        'mysql' => [
            'adapter' => 'mysql',
            'pdo' =>  new DatabaseConnection(),
        ]
    ],
];
```

Or the other way: set directly implementation of Phoenix\Database\Adapter\AdapterInterface in config.

I would like to use only classes or implementations of interfaces which are already defined in Phoenix.

**Comment by MekDrop** [comment]
Yeah I'm thinking about same. 

Thanks for ideas!

I think I'll try a bit later today to make a pull request for this change.

**Comment by MekDrop** [comment]
I thought about better example why I need to use prefixes. and I have remembered that WordPress has very similar thing that I would like to achieve.  

Here is long description how database there works -> https://multilingualpress.org/docs/wordpress-multisite-database-tables/. But in short there are some tables that are shared between sites and there are some tables that has same name, structure but different prefix for different web sites.

I have idea how to implement if you have multiple databases with phoenix - just just automatically generated different environments for different sites, but this is not good solution for many cheap sharing hosting providers (yeah still in 2020!) 

I think two code clean solutions are possible for solve this:
1) add possibility to specify table prefix for environment
2) add possibility to use container for all migrations at least for constructors and remove `final` keywords (I can't see any reasons why this must be final) for many primary migration functions, so these methods could be extended with needed functionality with help of existing services from the project
3) like 2 but don't use container; change 
```php
$migrations[$migrationIdentifier] =  new $className($this->adapter);
``` 
 into something like this:
```php
$refl = new \ReflectionClass($className);
$migrations[$migrationIdentifier]= $refl->newInstanceArgs(
    array_map(
        function ($parameter) {
          $type = $parameter->getType()->getName();    
          return new $type();
        },
        $refl->getConstructor()->getParameters()
    )
);
```
in `\Phoenix\Migration\Manager`

So some very primitive autowiring would work and that would help to use some functionality.

1 variant would be easy to use but can be complicated to implement. 3 would be easiest way to implement.

**Comment by lulco** [comment]
Hi, as I know, WordPress creates some tables for each user, but it will not work with phoenix, because each migration is executed only once, you can't run it with another prefix or so.

**Comment by MekDrop** [comment]
I think this will be solved if everything ok with #208 pull request

**Comment by lulco** [comment]
Hi @MekDrop I just implemented simple dependency injection to migrations as you suggested. Maybe it will help you to solve some problems. Check master branch for this feature.

Registration of classes for DI is in config array:
```
return [
    'migration_dirs' => [
        'phoenix' => __DIR__ . '/../phoenix',
    ],
    'environments' => [
       ...
    ],
    'dependencies' => [
        UuidFactoryInterface::class => new UuidFactory(),
    ],
];
```

Key is the type which is used in __construct of Migration, value is the implementation of the type.
